# ML Zoomcamp - Homework #1

http://mlzoomcamp.com/

- All the comments were made in portuguese to contribute with people that do not have too much familiarity with english.
- Todos os comentários foram feitos em português para contribuir com o aprendizado de pessoas que não tem tanta familiaridade com a língua inglesa.

- I also want to produce some content in portuguese to learners in Brazil and help to divulge this course.
- Também desejo produzir conteúdo em português para estudantes no Brasil e ajudar a divulgar esse curso!

In [2]:
# importando bibliotecas 
import numpy as np # matrizes
import pandas as pd # tratamento do conjunto de dados

## Questão 1

- Qual a versão do pandas instalada?

In [3]:
pd.__version__ # versão corrente da biblioteca numpy

'1.4.3'

In [4]:
# caminho e carregamento do conjunto de dados que será utilizado no trabalho
path = 'https://raw.githubusercontent.com/alexeygrigorev/datasets/master/housing.csv' 
df = pd.read_csv(path)

## Questão 2

- Qual o número de colunas no conjunto de dados?

In [5]:
# formato dos dados em termos de número de linhas (amostras) e número de colunas (variáveis, fatores)
df.shape

(20640, 10)

In [6]:
df.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   

   population  households  median_income  median_house_value ocean_proximity  
0       322.0       126.0         8.3252            452600.0        NEAR BAY  
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY  
2       496.0       177.0         7.2574            352100.0        NEAR BAY  
3       558.0       219.0         5.6431            341300.0        NEAR BAY  
4       565.0       259.0         3.8462            342200.0        NEAR BAY

## Questão 3 

- Existem colunas com dados faltando? Quais?

In [8]:
df.isnull().sum()

longitude               0
latitude                0
housing_median_age      0
total_rooms             0
total_bedrooms        207
population              0
households              0
median_income           0
median_house_value      0
ocean_proximity         0
dtype: int64

#### Resposta: "Total Bedrooms" tem 207 dados faltando.

## Questão 4

- Quantos valores únicos existem na coluna 'ocean proximity'?

In [12]:
df['ocean_proximity'].unique().shape

(5,)

In [15]:
# Quais são os valores únicos?
df['ocean_proximity'].unique()

array(['NEAR BAY', '<1H OCEAN', 'INLAND', 'NEAR OCEAN', 'ISLAND'],
      dtype=object)

## Questão 5

- Qual o valor médio da coluna "median_house_value" para as casas próximas a baía?

In [24]:
valor_medio = round(df[df['ocean_proximity'] == 'NEAR BAY']['median_house_value'].mean(),2)
print(f'O valor médio de casas próximas a baía é de: ${valor_medio}.')

O valor médio de casas próximas a baía é de: $259212.31.


## Questão 6

- Encontre o valor da média da coluna "total_bedrooms" no conjunto de dados.
- Use o método "fillna" para preencher os valores ausentes na coluna "total_bedrooms" com o valor da média encontrado na etapa anterior.
- Por fim, calcule o valor da média da coluna "total_bedrooms" mais uma vez.

Ele mudou?

In [25]:
# valor médio antes de preencher os valores nulos
df['total_bedrooms'].mean()

537.8705525375618

In [26]:
# preenchendo os valores nulos com os valores da média da respectiva coluna
df['total_bedrooms'] = df['total_bedrooms'].fillna(value=df['total_bedrooms'].mean())

In [27]:
# conferindo a alteração no número de valores nulos
df['total_bedrooms'].isnull().sum()

0

In [28]:
# observando se o valor da média permaneceu o mesmo após a alteração dos valores nulos
df['total_bedrooms'].mean()

537.8705525375617

- O valor permaneceu o mesmo após o tratamento dos valores ausentes. 

## Questão 7

- Selecione todos os valores que sejam "Ilhas" (Island);
- Selecione apenas as colunas "housing_median_age", "total_rooms", "total_bedrooms";
- Obtenha a matriz NumPy subjacente. Vamos chamá-la de X.
- Calcule a multiplicação matriz-matriz entre a transposição de X e X. Para obter a transposição, use X.T. Vamos chamar o resultado de XTX.
- Inverta XTX.
- Crie uma matriz y com valores [950, 1300, 800, 1000, 1300].
- Multiplique o inverso de XTX com a transposição de X e, em seguida, multiplique o resultado por y. Chame o resultado de w.

Qual é o valor do último elemento de w?

In [30]:
# selecionando todos os valores que sejam ISLAND e apenas as 3 colunas citadas
df_islands = df[df['ocean_proximity'] == 'ISLAND'][['housing_median_age', 'total_rooms', 'total_bedrooms']]
# resultado
df_islands

housing_median_age  total_rooms  total_bedrooms
8314                27.0       1675.0           521.0
8315                52.0       2359.0           591.0
8316                52.0       2127.0           512.0
8317                52.0        996.0           264.0
8318                29.0        716.0           214.0

In [31]:
X = np.array(df_islands) # criação de uma matriz a partir do conjunto de dados anterior
X

array([[  27., 1675.,  521.],
       [  52., 2359.,  591.],
       [  52., 2127.,  512.],
       [  52.,  996.,  264.],
       [  29.,  716.,  214.]])

In [32]:
XTX = X.T.dot(X) # produto entre a matriz transposta de X e a própria matriz X
XTX

array([[9.6820000e+03, 3.5105300e+05, 9.1357000e+04],
       [3.5105300e+05, 1.4399307e+07, 3.7720360e+06],
       [9.1357000e+04, 3.7720360e+06, 9.9835800e+05]])

In [33]:
XTX_inv = np.linalg.inv(XTX) # calculando a matriz inversa de XTX

In [34]:
XTX_inv

array([[ 9.19403586e-04, -3.66412216e-05,  5.43072261e-05],
       [-3.66412216e-05,  8.23303633e-06, -2.77534485e-05],
       [ 5.43072261e-05, -2.77534485e-05,  1.00891325e-04]])

In [35]:
y = np.array([950, 1300, 800, 1000, 1300]) 

In [36]:
z = XTX_inv.dot(X.T) # armazenando a multiplicação da matriz inversa de XTX com a transposta de X em uma variável "z"
w = z.dot(y) # calculando o produto da matriz z com o vetor y para obter w

In [37]:
w

array([23.12330961, -1.48124183,  5.69922946])

In [38]:
w[-1] # último elemento do vetor "w"

5.699229455065586